plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java

```diff
@@ -2226,17 +2227,33 @@ int getReservedCpuMHZ(VirtualMachineTO vmSpec) {
         return new String[] {datastoreDiskPath};
     }
 
-    // Pair<internal CS name, vCenter display name>
-    private Pair<String, String> composeVmNames(VirtualMachineTO vmSpec) {
-        String vmInternalCSName = vmSpec.getName();
-        String vmNameOnVcenter = vmSpec.getName();
-        if (_instanceNameFlag && vmSpec.getHostName() != null) {
-            vmNameOnVcenter = vmSpec.getHostName();
+    /**
+     * This method generates VM name for Vcenter and Cloudstack( when Hypervisor is VMware).
+     * It generates VM name according to  _instanceNameFlag setting.
+     *
+     * @param VirtualMachineTO
+     *        vmSpec
+     * @return Pair<internal CS name, vCenter display name>. A pair which contain 'internal CS name'  and
+     *         'vCenter display name'(vCenter knows VM by this name).
+     **/
+    protected Pair<String, String> composeVmNames(VirtualMachineTO vmSpec) {
+
+        String vmInternalCSName = "";
+        String vmNameOnVcenter  = "";
+        if(vmSpec != null)
+        {
+            vmInternalCSName = vmNameOnVcenter = vmSpec.getName();
+            if (_instanceNameFlag == true && vmSpec.getType() == VirtualMachine.Type.User) {
+                String[] tokens = vmInternalCSName.split("-");
+                if(tokens.length<3) throw new InvalidParameterValueException("Invalid vmInternalCSName format: "+vmInternalCSName+" expected format : *-*-* (e.g -> i-x-y)"); // vmInternalCSName has format i-x-y-<instance.name>
+                vmNameOnVcenter = String.format("%s-%s-%s-%s", tokens[0], tokens[1], tokens[2], vmSpec.getHostName());
```

Comment:
>As per previous code, vmNameOnVcenter = vmSpec.getHostName(), which will introduce backward compatibility issues for older 4.3 environments. Is there any code that handles for such old VMs.

Yes, vm.instancename.flag will handle backward compatibility issue. 

Case (vm.instancename.flag = True):
User can create the VM with same name form different account. 
Case (vm.instancename.flag = False):
User can not create the VM with same name form different account.
Note - But in both the cases, VM with same name are not allowed on a particular network. And only for the User VM's name format will be like "I-<user-id>-<Display-Name>" if "vm.instancename.flag = True". It will not cause any impact on other VM names( like Router VM).